Deprecation warning

brunobeltran opened the pull request:
## PR Summary

We are currently unable to unintrusively deprecate class-level attributes (because we subclass UserWarning and not DeprecationWarning). Based on a previous dev call, we decided the solution was to "do it anyway". This led to a couple of issues, e.g. #19080 and #19839 and #19850. 

The history of the decision to not subclass DeprecationWarning has to do with a decision from core Python in the 2.x days to not show DeprecationWarnings to users. However, there is now a more sophisticated filter in place (see https://www.python.org/dev/peps/pep-0565/). 

Users that want to see MatplotlibDeprecationWarning in their CI will now have to `export PYTHONWARNING=d`, launch python with `-Wa`, or use warning filters to enable DeprecationWarnings (or MatplotlibDeprecationWarning specifically, depending on their needs).


## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

anntzer commented:
https://docs.python.org/3/library/exceptions.html#DeprecationWarning or https://docs.python.org/3/library/exceptions.html#FutureWarning?

brunobeltran commented:
CI failures seem unrelated.

jklymak reviewed (approved):
Reading at @anntzer link I still think this is the right deprecation here

timhoffm commented:
Behavior for showing DeprecationWarnings according to https://www.python.org/dev/peps/pep-0565/

Warning shows, when using deprecated functions

- in interactive prompts, including notebooks
- in scripts
- in pytest runs

> Warnings will continue to be silenced by default for packaged code distributed as part of an importable module.

I failed to construct such a case :eyes: 

This still warns:

```
myscript.py  -- calls -->  mylib.plot() -- calls --> a deprecated Matplotlib function
```
where mylib is next to myscript. Maybe this is only suppressed if mylib is in site-packages?

Anyway given the above warning cases, I think it's safe enough to switch to `DeprecationWarning`.

timhoffm commented:
Does this need an API change note?

jklymak commented:
Definitely needs an API note!

brunobeltran commented:
Squashed and added an API note. 

Unsure where to dump all the information for most readability? Maybe a blog post on best practices for dealing with warnings for users/CI/library authors and would be more effective than putting some of this information in the docs?
